### PR TITLE
Restore same-symbol highlighting

### PR DESCRIPTION
**Pull request by erikrose**
Fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1123877

While we're at it, clean up the API and implementation. regions() and refs() now return Region and Ref classes which will be extensible in the future as we add things (like we've added `hover` and `qualname_hash` to refs).

### CODE REVIEW

File: dxr/indexers.py [hunk 1]
```diff
@@ -406,6 +394,101 @@ def needles_by_line(self):
...
+
+    def __init__(self, css_class):
+        self.css_class = css_class
+
+    def es(self):
+        return self.css_class
+
+    @classmethod
+    def es_to_triple(cls, es_region):
+        """Convert ES-dwelling refs representation to a (start, end, Region)
```
Comment (pelmers) [inline dxr/indexers.py]: "refs repr" should be "region representation" here.

File: dxr/lines.py [hunk 2]
```diff
@@ -331,7 +296,7 @@ def finished_tags(lines, refs, regions):
     # Plugins return unicode offsets, not byte ones.
 
     # Get start and endpoints of intervals:
-    tags = list(tag_boundaries(refs, regions))
+    tags = list(tag_boundaries(list(chain(refs, regions))))
```
Comment (pelmers) [inline dxr/lines.py]: Is it necessary to call list on the chained sequence? (it looks like it's only iterated over once)

File: dxr/lines.py [hunk 3]
```diff
@@ -427,17 +374,9 @@ def segments(text, tags, bof_offset):
...
-                menu, hover = payload.payload
-                menu = cgi.escape(json.dumps(menu), True)
-                if hover:
-                    title = ' title="' + cgi.escape(hover, True) + '"'
-                else:
-                    title = ''
-                yield u'<a data-menu="%s"%s>' % (menu, title)
+                yield payload.closer()
+            else:
+                yield payload.opener()
```
Comment (pelmers) [inline dxr/lines.py]: This reads really nicely now!

File: dxr/lines.py (hunk 2 above)
Comment (erikrose) [inline dxr/lines.py]: Good catch! That was just for debugging.

File: dxr/lines.py [hunk 4]
```diff
@@ -340,7 +305,7 @@ def finished_tags(lines, refs, regions):
     tags.sort(key=nesting_order)  # balanced_tags undoes this, but we tolerate
                                   # that in html_lines().
     remove_overlapping_refs(tags)
-    return balanced_tags(tags)
+    return list(balanced_tags(tags))
```
Comment (pelmers) [inline dxr/lines.py]: Also making a list here, which doesn't seem necessary

**Comment by pelmers** [comment]
r+

**Pull request merged**